x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/CategorizeTextAggregator.java

```diff
@@ -0,0 +1,208 @@
+import org.elasticsearch.search.lookup.SourceLookup;
+import org.elasticsearch.xpack.core.ml.job.config.CategorizationAnalyzerConfig;
+import org.elasticsearch.xpack.ml.job.categorization.CategorizationAnalyzer;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+public class CategorizeTextAggregator extends DeferableBucketAggregator {
+
+    private final TermsAggregator.BucketCountThresholds bucketCountThresholds;
+    private final SourceLookup sourceLookup;
+    private final BigArrays bigArrays;
+    private final MappedFieldType fieldType;
+    private final CategorizationAnalyzer analyzer;
+    private final String sourceFieldName;
+    private ObjectArray<CategorizationTokenTree> categorizers;
+    private final int maxChildren;
+    private final int maxDepth;
+    private final double similarityThreshold;
+    private final LongKeyedBucketOrds bucketOrds;
+
+    protected CategorizeTextAggregator(
+        String name,
+        AggregatorFactories factories,
+        AggregationContext context,
+        Aggregator parent,
+        String sourceFieldName,
+        MappedFieldType fieldType,
+        TermsAggregator.BucketCountThresholds bucketCountThresholds,
+        int maxChildren,
+        int maxDepth,
+        double similarityThreshold,
+        List<String> categorizationFilters,
+        Map<String, Object> metadata
+    ) throws IOException {
+        super(name, factories, context, parent, metadata);
+        this.sourceLookup = context.lookup().source();
+        this.sourceFieldName = sourceFieldName;
+        this.fieldType = fieldType;
+        CategorizationAnalyzerConfig categorizationAnalyzerConfig = CategorizationAnalyzerConfig.buildStandardCategorizationAnalyzer(
+            categorizationFilters
+        );
+        this.analyzer = new CategorizationAnalyzer(context.getAnalysisRegistry(), categorizationAnalyzerConfig);
+        this.bigArrays = context.bigArrays();
+        this.categorizers = bigArrays().newObjectArray(1);
+        this.maxChildren = maxChildren;
+        this.maxDepth = maxDepth;
+        this.similarityThreshold = similarityThreshold;
+        this.bucketOrds = LongKeyedBucketOrds.build(bigArrays(), CardinalityUpperBound.MANY);
+        this.bucketCountThresholds = bucketCountThresholds;
+    }
+
+    @Override
+    protected void doClose() {
+        super.doClose();
+        this.analyzer.close();
+    }
+
+    @Override
+    public InternalAggregation[] buildAggregations(long[] ordsToCollect) throws IOException {
+        InternalCategorizationAggregation.Bucket[][] topBucketsPerOrd =
+            new InternalCategorizationAggregation.Bucket[ordsToCollect.length][];
+        for (int ordIdx = 0; ordIdx < ordsToCollect.length; ordIdx++) {
+            int size = (int) Math.min(bucketOrds.size(), bucketCountThresholds.getShardSize());
```

Comment:
`bucketOrds.size()` is the number of buckets in the whole structure thing for all the `owningBucketOrd`s. You may want `bucketsInOrd` here. That's linear with the number of buckets but it hasn't shown to be a problem. Yet.

Reply:
Or maybe your tree can tell you how many buckets it has?